### PR TITLE
Multitask Learning

### PR DESCRIPTION
**Pull request by ben-heil**
This PR implements the training logic and models for multitask learning. The training logic in `train_multitask` is kind of messy, my next PR will refactor large portions of `classifier.py` to fix that

Changes:
- Created the `train_multitask` function. This function is based of `train_with_erm`, but trains a shared representation and classifiers on multiple diseases instead of training on a single disease.
- Added a parameter to `classifier.py` in the main function with accompanying logic that runs multitask learning
- Created the `ExpressionRepresentation` and `MultitaskHead` model classes in `model.py`. The latter is equivalent to the output node of a normal neural net, doing binary classification on whether the input corresponds to a given disease. The former is shared between all task heads and learns a representation of the gene expression that works for all diseases.
- Added the `get_gene_count` function to util.py to count the number of genes in the gene file
- Upgraded `add_study_ids_to_results` to allow for groups of studies to be added incrementally instead of all at one time

### CODE REVIEW

File: whistl/classifier.py [hunk 1]
```diff
@@ -17,6 +19,9 @@
 import model
 import util
 
+# import pudb
```
Comment (jjc2718) [inline whistl/classifier.py]: You can probably remove this

File: whistl/classifier.py [hunk 2]
```diff
@@ -350,6 +354,181 @@ def train_with_erm(classifier, map_file, train_dirs, tune_dirs, gene_file, num_e
...
+        'healthy' should not be in this list, as most studies will contain healthy samples
+    map_file: string or Path
+        The file created by label_samples.py to be used to match samples to labels
+    data_dirs: list of strs or list of Paths
+        The directories containing gene expression data for various experiments
+    gene_file: string or Path
+        The path to the file containing the list of genes to use in the model
+    num_epochs: int
+        The number of times the model should be trained on all the data
+    loss_scaling_factor: float
```
Comment (jjc2718) [inline whistl/classifier.py]: Are you using this here? I don't see it anywhere in the function.

File: whistl/classifier.py [hunk 3]
```diff
@@ -350,6 +354,181 @@ def train_with_erm(classifier, map_file, train_dirs, tune_dirs, gene_file, num_e
...
+                results['train_loss'].append(train_loss / len(train_dataset))
+                results['tune_loss'].append(tune_loss / len(tune_dataset))
+                results['train_acc'].append(train_accuracy)
+                results['tune_acc'].append(tune_accuracy)
+        except Exception as e:
+            # Print error
+            logger.error(e, exc_info=True)
+        finally:
+            if save_file_dir is not None:
+                representation_file = os.path.join(save_file_dir, 'erm_representation.pkl')
```
Comment (jjc2718) [inline whistl/classifier.py:513]: Maybe I'm just misunderstanding what you're trying to do here, but is there any reason why you're saving the "best" head model above on line 499 (essentially doing early stopping for the head model), but you're saving the representation from the last epoch here even if it doesn't have the best loss?

It seems to me like you'd also want to save the representation with the best loss (i.e. the representation that corresponds to the saved head model), but I could be confused about the purpose of this.

File: whistl/classifier.py [hunk 4]
```diff
@@ -350,6 +354,181 @@ def train_with_erm(classifier, map_file, train_dirs, tune_dirs, gene_file, num_e
...
+        A dictionary containing lists tracking different loss metrics across epochs
+    '''
+    sample_to_label = util.parse_map_file(map_file)
+
+    num_genes = util.get_gene_count(gene_file)
+    # Initialize the representation portion of the model
+    representation = representation(num_genes).double().to(device)
+
+    task_to_results = {}
+    for task in labels:
```
Comment (jjc2718) [inline whistl/classifier.py:395]: Will the order of the tasks matter here? My understanding of multitask learning was that all of the tasks are trained jointly, so for each training epoch you'd calculate the loss for each task, then add them to get the total loss - are you taking a different approach and I'm just thinking of a particular sub-category of multitask learning?

**Review by jjc2718 — APPROVED** [review]
Looks like a good starting point! Just a few clarification questions and small semantic things.

File: whistl/classifier.py (hunk 1 above)
Comment (ben-heil) [reply whistl/classifier.py]: Yup, will do

File: whistl/classifier.py (hunk 2 above)
Comment (ben-heil) [reply whistl/classifier.py]: Good catch, I pulled the training logic from `train_with_erm` but I must have grabbed the parameters from `train_with_irm`.

File: whistl/classifier.py (hunk 3 above)
Comment (ben-heil) [reply whistl/classifier.py:513]: I don't really know what works best yet. The purpose of this PR was to sketch out the base function to use as reference so I can benchmark different methods later. Currently each task is being trained in its entirety before moving onto the next, but it could be done a number of different ways. In a future PR I plan on comparing
- The current method
- Train only one batch per task at a time
- Do joint training (easier training code, probably better performance, but harder to swap out task heads further down the line)

I think the correct way to do it would be to train the representation, without keeping the heads, then freeze the representation weights and retrain the each head on the optimal representation.

All of that is to say that you are correct that the current method doesn't make much sense. I'm alright with that because it's more of a prototype than a method I intend to use in analyses

File: whistl/classifier.py (hunk 4 above)
Comment (ben-heil) [reply whistl/classifier.py:395]: It's possible to train tasks jointly or separately. I did it separately first because I'm more familiar with joint training and wanted to make sure that swapping heads was working correctly
